src/CoreWCF.Kafka/src/CoreWCF/Channels/KafkaTransportPump.cs

```diff
@@ -0,0 +1,246 @@
+            (KafkaDeliverySemantics.AtLeastOnce, { EnableAutoCommit: true, EnableAutoOffsetStore: false } ) => s_atLeastOnceBatchCommitConfigValues,
+            (KafkaDeliverySemantics.AtLeastOnce, { EnableAutoCommit: false, EnableAutoOffsetStore: null }) => s_atLeastOncePerMessageCommitConfigValues,
+            _ => throw new NotSupportedException(string.Format(SR.InvalidKafkaConfiguration, kafkaDeliverySemantics, consumerConfig.EnableAutoCommit, consumerConfig.EnableAutoOffsetStore))
+        };
+
+    private void OnLog(IConsumer<Null, byte[]> consumer, LogMessage logMessage)
+    {
+        const string format = "{0}:{1}";
+        switch (logMessage.Level)
+        {
+            case SyslogLevel.Debug:
+                _logger.LogDebug(format, logMessage.Name, logMessage.Message);
+                break;
+            case SyslogLevel.Notice:
+            case SyslogLevel.Info:
+                _logger.LogInformation(format, logMessage.Name, logMessage.Message);
+                break;
+            case SyslogLevel.Warning:
+                _logger.LogWarning(format, logMessage.Name, logMessage.Message);
+                break;
+            case SyslogLevel.Error:
+                _logger.LogError(format, logMessage.Name, logMessage.Message);
+                break;
+            case SyslogLevel.Alert:
+            case SyslogLevel.Critical:
+            case SyslogLevel.Emergency:
+                _logger.LogCritical(format, logMessage.Name, logMessage.Message);
+                break;
+            default:
+                throw new ArgumentOutOfRangeException(nameof(logMessage.Level));
+        }
+    }
+
+    private void OnError(IConsumer<Null, byte[]> consumer, Error error)
+    {
+        if (error.IsFatal)
+        {
+            _cts.Cancel();
+        }
+    }
+
+    private Task OnConsumeMessage(ConsumeResult<Null, byte[]> consumeResult,
+        QueueTransportContext queueTransportContext)
+    {
+        var receiveContext = new KafkaReceiveContext(consumeResult, this);
+        IncrementReceiveContextCount();
+        var context = new QueueMessageContext
+        {
+            QueueMessageReader = PipeReader.Create(new ReadOnlySequence<byte>(consumeResult.Message.Value)),
+            LocalAddress = new EndpointAddress(queueTransportContext.ServiceDispatcher.BaseAddress),
+            QueueTransportContext = queueTransportContext,
+            ReceiveContext = receiveContext
+        };
+
+        return queueTransportContext.QueueMessageDispatcher(context);;
+    }
+
+    private void IncrementReceiveContextCount()
+    {
+        _receiveContextCountdownEvent.AddCount();
+    }
+
+    internal void DecrementReceiveContextCount()
+    {
+        _receiveContextCountdownEvent.Signal();
+    }
+
+    public void Dispose()
+    {
+        if (TransportBindingElement.ErrorHandlingStrategy == KafkaErrorHandlingStrategy.DeadLetterQueue)
+        {
+            Producer.Dispose();
+        }
```

Comment:
If Dispose is called before StartPumpAsync, this will throw NRE. It might just be simpler and safer to call `Producer?.Dispose(); Producer = null;` You need not protect against multiple Dispose calls throwing. What happens if `Consumer.Dispose()` is called multiple times? Does it throw or silently ignore it? 
What happens if Dispose() is called before StopPumpAsync?

Reply:
it should be ok now i added null guards and a lock